tests/basic.rs

```diff
@@ -1,486 +1,860 @@
-    assert!(
-        !actual.display().to_string().contains("/./"),
-        "'{}' should not contain a CurDir component",
-        actual.display()
-    );
-}
+        fn with_entry_mut(&mut self, path: impl AsRef<Path>) -> Option<&mut DirectoryEntry> {
+            let mut current_entry = &mut self.root_dir;
+            let mut components = path.as_ref().components().peekable();
+
+            while let Some(component) = components.next() {
+                match component {
+                    Component::RootDir => {
+                        let is_last = components.peek().is_none();
+                        if is_last {
+                            return Some(current_entry);
+                        }
+                    }
+                    Component::Normal(os_str) => {
+                        let is_last = components.peek().is_none();
+                        let entry = current_entry
+                            .unwrap_directory_mut()
+                            .entries
+                            .get_mut(os_str)?;
+
+                        if is_last {
+                            return Some(entry);
+                        }
+
+                        current_entry = entry;
+                    }
+                    Component::CurDir | Component::ParentDir | Component::Prefix(_) => todo!(),
+                }
+            }
+            None
+        }
 
-#[test]
-#[cfg(unix)]
-fn test_which_non_executable() {
-    // Shouldn't return non-executable files.
-    let f = TestFixture::new();
-    f.touch("b/another", "").unwrap();
-    assert!(_which(&f, "another").is_err());
-}
+        fn get_entry(&self, path: &Path) -> Option<&DirectoryEntry> {
+            let mut current_entry = &self.root_dir;
+            let mut components = path.components().peekable();
+
+            while let Some(component) = components.next() {
+                match component {
+                    Component::RootDir => continue,
+                    Component::Normal(os_str) => {
+                        let entry = current_entry.unwrap_directory().entries.get(os_str)?;
+                        if components.peek().is_none() {
+                            return Some(entry);
+                        } else {
+                            current_entry = entry;
+                        }
+                    }
+                    Component::CurDir | Component::ParentDir | Component::Prefix(_) => todo!(),
+                }
+            }
+
+            unreachable!()
+        }
 
-#[test]
-#[cfg(unix)]
-fn test_which_absolute_non_executable() {
-    // Shouldn't return non-executable files, even if given an absolute path.
-    let f = TestFixture::new();
-    let b = f.touch("b/another", "").unwrap();
-    assert!(_which(&f, b).is_err());
-}
+        fn get_entry_follow_symlink(&self, path: &Path) -> Option<&DirectoryEntry> {
+            let mut current_path = path.to_path_buf();
+            let mut seen = std::collections::HashSet::new();
```

Comment:
Here's one

Reply:
Thanks, I searched for `HashSet<` and didn't find it because of that.